### ISSUE TITLE
OpenSSL Not Closing Resource Deep in OkHTTP

### ISSUE THREAD

**Issue by jophde** [square/okhttp#1854]
I work at Optimizely on our SDKs.  We depend on okhttp for our Android SDK.  A customer reported an issue to us that occurs when they run their app in strict mode.  They are running the app in strict mode to check for issues with our SDK.  When they do the app crashes on startup.

They setup strict in their launch Activity's onCreate method like this:

```
StrictMode.setThreadPolicy(new StrictMode.ThreadPolicy.Builder()
                .detectAll()   // or .detectAll() for all detectable problems
                .penaltyLog()
                .build());
StrictMode.setVmPolicy(new StrictMode.VmPolicy.Builder()
                .detectLeakedSqlLiteObjects()
                .detectLeakedClosableObjects()
                .penaltyLog()
                .penaltyDeath()
                .build());
```

This leads to the following stack trace after a crash:
```
09-15 03:27:26.799    7500-7509/? E/StrictMode﹕ A resource was acquired at attached stack trace but never released. See java.io.Closeable for information on avoiding resource leaks.
    java.lang.Throwable: Explicit termination method 'close' not called
            at dalvik.system.CloseGuard.open(CloseGuard.java:184)
            at com.android.org.conscrypt.OpenSSLSocketImpl.startHandshake(OpenSSLSocketImpl.java:278)
            at com.squareup.okhttp.internal.http.SocketConnector.connectTls(SocketConnector.java:103)
            at com.squareup.okhttp.Connection.connect(Connection.java:143)
            at com.squareup.okhttp.Connection.connectAndSetOwner(Connection.java:185)
            at com.squareup.okhttp.OkHttpClient$1.connectAndSetOwner(OkHttpClient.java:128)
            at com.squareup.okhttp.internal.http.HttpEngine.nextConnection(HttpEngine.java:341)
            at com.squareup.okhttp.internal.http.HttpEngine.connect(HttpEngine.java:330)
            at com.squareup.okhttp.internal.http.HttpEngine.sendRequest(HttpEngine.java:248)
            at com.squareup.okhttp.Call.getResponse(Call.java:273)
            at com.squareup.okhttp.Call$ApplicationInterceptorChain.proceed(Call.java:230)
            at com.squareup.okhttp.Call.getResponseWithInterceptorChain(Call.java:201)
            at com.squareup.okhttp.Call.execute(Call.java:81)
            at com.optimizely.Network.OptimizelyNetworkUtil.downloadFromUrl(OptimizelyNetworkUtil.java:68)
            at com.optimizely.Network.OptimizelyDownloader.downloadFromUrl(OptimizelyDownloader.java:188)
            at com.optimizely.Network.OptimizelyDownloader$DownloadTask.doInBackground(OptimizelyDownloader.java:150)
            at com.optimizely.Network.OptimizelyDownloader$DownloadTask.doInBackground(OptimizelyDownloader.java:129)
            at android.os.AsyncTask$2.call(AsyncTask.java:288)
            at java.util.concurrent.FutureTask.run(FutureTask.java:237)
            at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
            at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
            at java.lang.Thread.run(Thread.java:841)
```

I have reproduced this in our dev app for okhttp 2.4.0 and okhttp 2.5.0.  Seems like a close is missing from a finally somewhere.  Please let me know how I can help!

**Comment by swankjesse** [comment]
Can you rephrase this question as a test case?

**Comment by swankjesse** [comment]
Ie. I'd love to see a code sample that causes this stack trace, where the code calling into OkHttp calls `close()`.

**Comment by jophde** [comment]
Here is the code that calls into OkHttp:

```java
@NonNull
    public Pair<T, Integer> downloadFromUrl(@NonNull OkHttpClient client, @NonNull String url) {
        client.setConnectTimeout(networkTimeout, TimeUnit.MILLISECONDS);
        client.setRetryOnConnectionFailure(false);
        try {
            Request request = new Request.Builder()
                .url(url)
                .build();
            optimizely.verboseLog(OPTIMIZELY_NETWORK_UTIL,
                "Executing request %s", request.urlString());
            Response response = client.newCall(request).execute();
            if (response.isSuccessful()) {
                ResponseBody body = response.body();
                try {
                    if (body != null) {
                        return new Pair<>(transform.transform(body), -1);
                    } else {
                        return new Pair<>(null, ERROR_UNKNOWN);
                    }
                } finally {
                    if(body != null) {
                        body.close();
                    }
                }
            } else {
                return new Pair<>(null, ERROR_4XX);
            }
        } catch (IOException e) {
            optimizely.verboseLog(true, OPTIMIZELY_NETWORK_UTIL,
                "Unable to download file : %1$s", e.getMessage());
            return new Pair<>(null, ERROR_NETWORK);
        } catch (SecurityException e) {
            optimizely.verboseLog(true, OPTIMIZELY_NETWORK_UTIL,
                "Cannot download file, possibly INTERNET permission not available. " +
                    "Got exception: %1$s", e.getLocalizedMessage());
            return new Pair<>(null, ERROR_NO_PERMISSION);
        } catch (Exception e) {
            optimizely.verboseLog(true, OPTIMIZELY_NETWORK_UTIL,
                "Cannot download file, with exception: %1$s", e.getLocalizedMessage());
            return new Pair<>(null, ERROR_UNKNOWN);
        }
    }
```

Specifically, `Response response = client.newCall(request).execute();` is the last call in the trace before it enters into Okhttp.  Again, this only happens if the user is running with the strict mode settings above.  We have never seen this in production during normal usage.  I will work on a test case for you.

**Comment by swankjesse** [comment]
Maybe move the `try/finally` to close the body even if it is unsuccessful?

**Comment by jophde** [comment]
That did indeed fix it!  Thanks for your help.
